In [12]:
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.model_selection import train_test_split

### 1. Library Imports
- `pandas` is loaded for tabular data manipulation and loading CSV files.
- `sklearn.model_selection.train_test_split` is used to partition data into training and validation sets.
- `RandomForestClassifier` serves as our ensemble learning algorithm.
- Evaluation metrics (`accuracy_score`, `classification_report`, `confusion_matrix`) are imported to measure prediction performance across classes.

In [13]:
df= pd.read_csv("/content/Iris.csv")
df.head()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa


In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidthCm   150 non-null    float64
 3   PetalLengthCm  150 non-null    float64
 4   PetalWidthCm   150 non-null    float64
 5   Species        150 non-null    object 
dtypes: float64(4), int64(1), object(1)
memory usage: 7.2+ KB


### 2. Exploratory Data Inspection
- The Iris dataset consists of 150 rows and 6 columns: 4 numerical measurement features, 1 identifier column (`Id`), and 1 categorical target label (`Species`).
- `.info()` confirms there are no missing values (`non-null`) and verifies that measurement columns are numerical (`float64`).

In [15]:
df.isnull()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,False,False,False,False,False,False
1,False,False,False,False,False,False
2,False,False,False,False,False,False
3,False,False,False,False,False,False
4,False,False,False,False,False,False
...,...,...,...,...,...,...
145,False,False,False,False,False,False
146,False,False,False,False,False,False
147,False,False,False,False,False,False
148,False,False,False,False,False,False


In [16]:
df.describe()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm
count,150.000000,150.000000,150.000000,150.000000,150.000000
mean,75.500000,5.843333,3.054000,3.758667,1.198667
std,43.445368,0.828066,0.433594,1.764420,0.763161
min,1.000000,4.300000,2.000000,1.000000,0.100000
25%,38.250000,5.100000,2.800000,1.600000,0.300000
50%,75.500000,5.800000,3.000000,4.350000,1.300000
75%,112.750000,6.400000,3.300000,5.100000,1.800000
max,150.000000,7.900000,4.400000,6.900000,2.500000


In [17]:
df.columns

Index(['Id', 'SepalLengthCm', 'SepalWidthCm', 'PetalLengthCm', 'PetalWidthCm',
       'Species'],
      dtype='object')

### 3. Data Integrity & Summary
- `.isnull().sum()` confirms zero missing values across all features, requiring no imputation steps.
- `.describe()` reveals the range, mean, and standard deviation of flower measurements:
  - Petal length has high variance, ranging from 1.0 cm to 6.9 cm, making it a strong candidate for species separation.

In [18]:
X = df[["SepalLengthCm", "SepalWidthCm", "PetalLengthCm", "PetalWidthCm"]]
y = df["Species"]

In [19]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

print(f"\nTraining samples: {len(X_train)}")
print(f"Testing samples:  {len(X_test)}")


Training samples: 120
Testing samples:  30


### 4. Feature Selection & Dataset Splitting
- **Feature Selection:** The `Id` column is excluded to prevent data leakage, keeping only `SepalLengthCm`, `SepalWidthCm`, `PetalLengthCm`, and `PetalWidthCm` as feature matrix `X`.
- **Target:** `Species` is assigned to target vector `y`.
- **Stratified Split:** Data is split into 80% training (120 samples) and 20% testing (30 samples). Using `stratify=y` ensures that each class (Setosa, Versicolor, Virginica) maintains an equal 1:1:1 ratio in both train and test splits.

In [20]:
# 5. TRAIN THE CLASSIFIER
print("\nTraining the model...")
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)
print("Training complete!")


Training the model...
Training complete!


### 5. Model Training (Random Forest)
- A `RandomForestClassifier` with 100 decision trees (`n_estimators=100`) is instantiated and fitted on `X_train` and `y_train`.
- Random Forest is selected because it handles non-linear boundaries well and prevents overfitting through bagging and feature subsampling.

In [21]:
# 6. EVALUATE THE MODEL
print("\nEvaluating the model...")
# Predict on unseen test data
y_pred = model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"\nModel Accuracy: {accuracy * 100:.2f}%")

print("\n--- Classification Report ---")
print(classification_report(y_test, y_pred))

print("--- Confusion Matrix ---")
print(confusion_matrix(y_test, y_pred))


Evaluating the model...

Model Accuracy: 90.00%

--- Classification Report ---
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        10
Iris-versicolor       0.82      0.90      0.86        10
 Iris-virginica       0.89      0.80      0.84        10

       accuracy                           0.90        30
      macro avg       0.90      0.90      0.90        30
   weighted avg       0.90      0.90      0.90        30

--- Confusion Matrix ---
[[10  0  0]
 [ 0  9  1]
 [ 0  2  8]]


### 6. Model Evaluation & Performance Analysis
- **Accuracy:** The model achieves 90% accuracy on unseen test data.
- **Iris-setosa:** Perfect performance (Precision = 1.00, Recall = 1.00, F1 = 1.00), demonstrating complete separability.
- **Versicolor vs. Virginica:** Slight overlap exists due to similar petal and sepal measurement distributions in borderline cases.

In [22]:
# 7. PREDICT A NEW FLOWER
print("\nPredicting a new flower...")
# Example: [SepalLength, SepalWidth, PetalLength, PetalWidth]
sample_flower = pd.DataFrame([[5.1, 3.5, 1.4, 0.2]], columns=X.columns)
prediction = model.predict(sample_flower)
# sample_flower = [[5.1, 3.5, 1.4, 0.2]]
# prediction = model.predict(sample_flower)
print(f"\nPrediction for sample {sample_flower.values[0]}: {prediction[0]}")


Predicting a new flower...

Prediction for sample [5.1 3.5 1.4 0.2]: Iris-setosa


### 7. Inference / Real-Time Prediction
- Tested with a sample flower of dimensions `[5.1, 3.5, 1.4, 0.2]`.
- The trained classifier predicts `Iris-setosa`, aligning with the typical short petal dimensions of the Setosa class.